pkg/operators/wasm/syscalls.go

```diff
@@ -0,0 +1,168 @@
+	wapi "github.com/tetratelabs/wazero/api"
+
+	syscallhelpers "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/traceloop/syscall-helpers"
+)
+
+func (i *wasmOperatorInstance) addSyscallsDeclarationsFuncs(env wazero.HostModuleBuilder) {
+	exportFunction(env, "getSyscallName", i.getSyscallName,
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Syscall ID
+		[]wapi.ValueType{wapi.ValueTypeI64}, // Syscall Name
+	)
+
+	exportFunction(env, "getSyscallDeclaration", i.getSyscallDeclaration,
+		[]wapi.ValueType{
+			wapi.ValueTypeI64, // Syscall Name
+			wapi.ValueTypeI64, // Syscall Declaration pointer
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // SyscallDeclaration or Error
+	)
+}
+
+// getSyscallName returns the syscall name corresponding to the given ID.
+// Params:
+// - stack[0]
+// Return value:
+// - Syscall name on success, 0 on error
+func (i *wasmOperatorInstance) getSyscallName(ctx context.Context, m wapi.Module, stack []uint64) {
+	syscallID := uint16(stack[0])
+	syscallName := syscallhelpers.SyscallGetName(syscallID)
+
+	bufPtr, err := i.writeToGuestMemory(ctx, []byte(syscallName))
+	if err != nil {
+		i.logger.Warnf("getSyscallName: allocating guest memory for %s: %v", syscallName, err)
+		stack[0] = 0
+		return
+	}
+
+	stack[0] = bufPtr
+}
+
+// Keep in sync with wasmapi/go/syscall.go.
+const (
+	isPointerFlag = 1 << iota
+)
+
+// Keep in sync with wasmapi/go/syscall.go.
+type syscallParam struct {
+	name  [32]byte
+	flags uint32
+}
+
+// Keep in sync with wasmapi/go/syscall.go.
+type syscallDeclaration struct {
+	// landlock_create_ruleset() is one of the longest syscall name with 24
+	// characters, let's round up to 32 to be sure.
+	name     [32]byte
+	nrParams uint8
+	_        [3]byte
+	// syscalls have maximum 6 arguments:
+	// https://github.com/torvalds/linux/blob/7cb1b4663150/include/linux/syscalls.h#L231
+	params [6]syscallParam
+}
+
+// getSyscallDeclaration gets a syscall declaration.
+// Params:
+// - stack[0] is the name of the syscall (string encoded)
+// - stack[1] is a pointer to a syscallDeclaration structure used to store the result.
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) getSyscallDeclaration(ctx context.Context, m wapi.Module, stack []uint64) {
+	syscallNamePtr := stack[0]
+	syscallDeclPtr := stack[1]
+
+	syscallName, err := stringFromStack(m, syscallNamePtr)
+	if err != nil {
+		i.logger.Warnf("getSyscallDeclaration: reading string from stack: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	// This map can be big, so let's do it only once and if needed.
+	i.syscallsDeclarations, err = sync.OnceValues(func() (map[string]syscallhelpers.SyscallDeclaration, error) {
+		return syscallhelpers.GatherSyscallsDeclarations()
+	})()
+	if err != nil {
+		i.logger.Warnf("getSyscallDeclaration: gathering syscall declarations: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	// This can only occur if above sync.Once failed, so let's return as the stack
+	// return value was correctly set.
```

Comment:
Is it really needed? I suppose the return on L110 will handle the failure case, and we can trust the function to return something different than nil when err == nil.